### ISSUE TITLE
MacTryserver failing Telemetry Tests with Stable Browser (Desktop)

### ISSUE THREAD

**Issue by simonhatch** [catapult-project/catapult#3510]
https://build.chromium.org/p/tryserver.client.catapult/builders/Catapult%20Mac%20Tryserver

**Comment by simonhatch** [comment]
Only recent commit that's mac specific is: https://github.com/catapult-project/catapult/commit/07274b7a0b08880de3427d7117515a2b1440f1b6

@chiniforooshan @kainino0x

**Comment by kainino0x** [comment]
@simonhatch My patch is mostly not mac specific, only the JSON entry is mac specific and I don't think it should be triggered at all yet. If a revert fixes it that would be great. I am digging through these logs to see what's going on there.

**Comment by simonhatch** [comment]
Awesome thanks for looking! My bad, just looked at the title and saw "Mac: ...", going to try repro on laptop.

**Comment by simonhatch** [comment]
The revert passes: https://build.chromium.org/p/tryserver.client.catapult/builders/Catapult%20Mac%20Tryserver/builds/7282

I'm doing a try on a fake cl with no changes just to confirm it's still broken

**Comment by simonhatch** [comment]
Repro'd this locally, it's stuck on asking for keychain access.

**Comment by kainino0x** [comment]
No problem, thank you for finding this. (Just to make sure - it does seem to be my patch, right?)

I don't know how to search these logs effectively - but I think this is the new error:

```
2017-04-20 13:50:22.344 Python[39500:228254] *** Terminating app due to uncaught exception 'NSRangeException', reason: '*** -[__NSArrayM objectAtIndex:]: index 0 beyond bounds for empty array'
```

I have no clue what's causing this, but I think I did see it briefly on previous versions of my patch. It seemed to go away permanently after that so I assumed it was unrelated to my patch at the time.

**Comment by simonhatch** [comment]
Yeah does seem to be your patch, my empty patch that I ran through the tryservers is stuck on the mac try server: https://build.chromium.org/p/tryserver.client.catapult/builders/Catapult%20Mac%20Tryserver/builds/7283

but the one I posted which reverted that CL passed easily.

**Comment by kainino0x** [comment]
Yes, sorry for the delay. We should revert it.

**Comment by kainino0x** [comment]
Newly failing tests (all of these weren't run without numpy/cv2):

* `telemetry.internal.image_processing.screen_finder_unittest.ScreenFinderTest.testBasicFunctionality`
    Seem flaky without `-j1` (i.e. if running more than one job at a time).

* `telemetry.internal.image_processing.video_file_frame_generator_unittest.VideoFileFrameGeneratorTest.testVideoFileFrameGenerator{Success,Failure,SkipFrames}`
    Also seems flaky without `-j1`

* `telemetry.internal.image_processing.video_file_frame_generator_unittest.VideoFileFrameGeneratorTest.testVideoFileFrameGenerator{Failure,SkipFrames}`
    These failures hit a crash inside cv2:
    `Terminating app due to uncaught exception 'NSRangeException', reason: '*** -[__NSArrayM objectAtIndex:]: index 0 beyond bounds for empty array'`
    (Side note: these tests never say "failed unexpectedly" in the logs above, but they do on my computer.)

I'm adding `@Disabled('mac`)` for these for now.

**Comment by nedn** [comment]
Oh, for those cases that pass with -j1, can you add @decorators.Isolated()?

**Comment by kainino0x** [comment]
Actually, changing all of them to `@Disabled('all')` because they also fail on my Linux machine.

On Linux, `testBasicFunctionality`, `testVideoFileFrameGeneratorSuccess`, and `testVideoFileFrameGeneratorSkipFrames` above fail even with `-j1`. They all fail in the same way:
`FrameReadError: Not enough frames in capture.`

`testVideoFileFrameGeneratorFailure` passes, but that's probably just because it expects failure.

**Comment by kainino0x** [comment]
@nedn I will see if I can tighten some of these to `@Disabled('linux') @Isolated()` or something like that.

**Comment by kainino0x** [comment]
* Mac: `telemetry.internal.image_processing.video_file_frame_generator_unittest.VideoFileFrameGeneratorTest.testVideoFileFrameGeneratorSuccess`
    actually seems to fail always, `0.0 != 33.367`.

**Comment by kainino0x** [comment]
* Mac: `telemetry.internal.image_processing.screen_finder_unittest.ScreenFinderTest.testBasicFunctionality`
    was found to fail even with `@decorators.Isolated()`. Disabling it as well.

**Comment by andrey-malets** [comment]
There are multiple issues involved here:

1. The `NSRangeException` stuff in logs is caused by a null pointer dereference in OpenCV code which uses Apple's AVFoundation APIs as a backend. The issue has already been fixed in master and 2.4 branches a week ago (here are "the fix" and "the fix for the fix" for 2.4: https://github.com/opencv/opencv/commit/f71bf21aaffec67739390c9fb58eafed68a35eae, https://github.com/opencv/opencv/commit/aae7621243d161d50e1c579925855a30b35bc8be). However, there were no recent releases from 2.4 and I doubt that the branch is still actively supported (the latest release was more than half a year ago). So an options are to update OpenCV to more recent version or leave the test causing the crash disabled (but read further).

2. The `0.0 != 33.367` (Mac) and `Not enough frames in capture` (Linux) issues should be caused by the same thing: OpenCV uploaded to CS via binary deps should have been built without FFmpeg support. Note that according to CMake files this support is enabled automatically when required libs (libavcodec etc) are found on a build machine (and disabled automatically if they are not), thus the person updating the libs should carefully examine CMake logs (or tune the options to make FFmpeg a requirement, but I haven't yet figured out the way to do this). Also note that FFmpeg proirity is more than that of AVFoundation on Mac, thus properly configured OpenCV build does not suffer from the first issue.

3. Finally, there is an import issue which causes all of the stuff above to flake: for example, the `image_processing` code which has dependencies on the `cv2` and `numpy` external modules gets imported transitively right from `run_tests.py`, even before `binary_manager` has a chance to kick in and download the right stuff. So there are 3 possible situations:
* The binary dependencies are already downloaded by the time we `run_tests`. The tests, as they were supposed to work, use stuff from `telemetry/third_party` and fail with all the effects described above.
* The binary dependencies are not downloaded yet (e.g. we execute the test suite on a clean build right after checkout) and the machine has no external dependencies present on their own (e.g. installed by the system package manager). Because import globs in `telemetry/telemetry/__init__.py` fail due to their "optional" nature, the tests are just... skipped being in some way "prepared" for this :)
* The binary dependencies are not downloaded yet, and the system has some versions of them already present in `sys.path`. In this case the tests will be using the system versions of libraries (and behave according to their buildflags, versions etc), even if `telemetry/third_party` versions get correctly downloaded and unpacked before test execution.

Obviously, if someone just runs tests a couple times on their machine, he may get repeated failures (case one) after "apparent success" (being a skip instead of a success, case two).

Now I'm thinking about the proper solution to 3, which may not be as simple.. Does someone have any ideas on this?

**Comment by nedn** [comment]
Thanks to @andrey-malets for your detective work!

I think the proper solution to 3 is we need to kick off `fetch_telemetry_binary_dependencies` to download all dependencies  before even invoking `run_tests`. We do this in Chromium repo through `gclient sync`, just hasn't done in catapult repo yet.

@anniesullie can you give the pointer to catapult recipe code? 

@kenrussell fyi

**Comment by anniesullie** [comment]
The catapult recipe is here: https://cs.chromium.org/chromium/build/scripts/slave/recipes/catapult.py

We use the `genearator_script` API which is documented here: https://github.com/luci/recipes-py/blob/master/recipe_modules/generator_script/api.py

Our code to generate steps is here: https://github.com/catapult-project/catapult/blob/master/catapult_build/build_steps.py

**Comment by nedn** [comment]
(3) from https://github.com/catapult-project/catapult/issues/3510#issuecomment-312696212  is fixed with the commit above, I think. We still need to address (1) & (2)

**Comment by kainino0x** [comment]
I agree with @nedn's solution to problem 3.

I will have to look into my build configuration for opencv on mac. Thanks Andrey for figuring out exactly what needs to change, it's super helpful!

We can update to a specific opencv revision beyond the 2.4.13.2 release. It is the latest 2.x release (unfortunately 3.x is incompatible) but I see no reason we need to be on a dotted release.

**Comment by andrey-malets** [comment]
> We can update to a specific opencv revision beyond the 2.4.13.2 release. It is the latest 2.x release (unfortunately 3.x is incompatible) but I see no reason we need to be on a dotted release.

Please note that it is not a requirement, 2.4.13.2 built with FFmpeg libs (I used the ones installed by macports) is sufficient.

**Comment by kainino0x** [comment]
Oh, right, I understand. Thanks for the clarification.
